contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,528 @@
+    }
+
+    function _verifyAdminsOfProxies(address admin) internal view {
+        _verifyProxyAdmin(_accountingOracle(), admin);
+        _verifyProxyAdmin(_locator, admin);
+        _verifyProxyAdmin(_stakingRouter(), admin);
+        _verifyProxyAdmin(_validatorsExitBusOracle(), admin);
+        _verifyProxyAdmin(_withdrawalQueue(), admin);
+        _verifyProxyAdmin(_withdrawalVault(), admin);
+    }
+
+    function _verifyProxyAdmin(IOssifiableProxy proxy, address admin) internal view {
+        if (proxy.proxy__getAdmin() != admin) revert WrongProxyAdmin(address(proxy));
+    }
+
+    function _verifyInitialProxyImplementations() internal view {
+        _verifyInitialImplementation(_accountingOracle());
+        _verifyInitialImplementation(_stakingRouter());
+        _verifyInitialImplementation(_validatorsExitBusOracle());
+        _verifyInitialImplementation(_withdrawalQueue());
+        _verifyInitialImplementation(_withdrawalVault());
+    }
+
+    function _verifyInitialImplementation(IOssifiableProxy proxy) internal view {
+        if (proxy.proxy__getImplementation() != _dummyImplementation) revert WrongInitialImplementation(address(proxy));
+    }
+
+    function _verifyOZAccessControlAdmins(address admin) internal view {
+        _verifySingleOZAdmin(_hashConsensusForAccountingOracle, admin);
+        _verifySingleOZAdmin(_hashConsensusForValidatorsExitBusOracle, admin);
+        _verifySingleOZAdmin(_burner(), admin);
+    }
+
+    function _verifySingleOZAdmin(IAccessControlEnumerable accessControlled, address admin) internal view {
+        if (accessControlled.getRoleMemberCount(DEFAULT_ADMIN_ROLE) != 1
+         || accessControlled.getRoleMember(DEFAULT_ADMIN_ROLE, 0) != admin
+        ) {
+            revert WrongOZAccessControlAdmin(address(accessControlled));
+        }
+    }
+
+    function _prepareWithdrawalVault() internal {
+        _withdrawalVault().proxy__upgradeTo(_withdrawalVaultImplementation);
+        _withdrawalVault().initialize();
+    }
+
+    function _prepareAccountingOracle() internal {
+        (, uint256 epochsPerFrame, ) = _hashConsensusForAccountingOracle.getFrameConfig();
+        uint256 lastLidoOracleCompletedEpochId = _lidoOracle().getLastCompletedEpochId();
+
+        // NB: HashConsensus.updateInitialEpoch must be called after AccountingOracle implementation is bound to proxy
+        _accountingOracle().proxy__upgradeTo(_accountingOracleImplementation);
+        _hashConsensusForAccountingOracle.updateInitialEpoch(lastLidoOracleCompletedEpochId + epochsPerFrame);
+        _accountingOracle().initialize(
+            address(this),
+            address(_hashConsensusForAccountingOracle),
+            _accountingOracleConsensusVersion
+        );
+    }
+
+    function _prepareValidatorsExitBus() internal {
+        _validatorsExitBusOracle().proxy__upgradeTo(_validatorsExitBusOracleImplementation);
+        _validatorsExitBusOracle().initialize(
```

Comment:
- This will determine how the oracles will tick, synchronously or not, and with what lag
- It's weird if the initial_epoch is earlier than the Exit Bus contract and the Lido contract itself